internal/service/dynamodb/global_secondary_index.go

```diff
@@ -0,0 +1,934 @@
+		)
+	}
+
+	action := &ddbtypes.CreateGlobalSecondaryIndexAction{
+		IndexName:  data.Name.ValueStringPointer(),
+		KeySchema:  keySchema,
+		Projection: projection,
+	}
+
+	if data.ReadCapacity.IsNull() || data.ReadCapacity.IsUnknown() {
+		data.ReadCapacity = types.Int64Value(0)
+	}
+
+	if data.WriteCapacity.IsNull() || data.WriteCapacity.IsUnknown() {
+		data.WriteCapacity = types.Int64Value(0)
+	}
+
+	billingMode := ddbtypes.BillingModeProvisioned
+	if table.BillingModeSummary != nil {
+		billingMode = table.BillingModeSummary.BillingMode
+	}
+
+	if billingMode == ddbtypes.BillingModeProvisioned {
+		rc := int(data.ReadCapacity.ValueInt64())
+		wc := int(data.WriteCapacity.ValueInt64())
+
+		if rc == 0 && table.ProvisionedThroughput != nil {
+			rc = int(aws.ToInt64(table.ProvisionedThroughput.ReadCapacityUnits))
+		}
+		if wc == 0 && table.ProvisionedThroughput != nil {
+			wc = int(aws.ToInt64(table.ProvisionedThroughput.WriteCapacityUnits))
+		}
+
+		provisionedThroughputData := map[string]any{
+			"read_capacity":  rc,
+			"write_capacity": wc,
+		}
+		action.ProvisionedThroughput = expandProvisionedThroughput(provisionedThroughputData, billingMode)
+	} else if !data.OnDemandThroughputs.IsNull() && !data.OnDemandThroughputs.IsUnknown() {
+		var odts []onDemandThroughputModel
+		response.Diagnostics.Append(data.OnDemandThroughputs.ElementsAs(ctx, &odts, false)...)
+		if len(odts) > 0 {
+			v := map[string]any{
+				"max_read_request_units":  int(odts[0].MaxReadRequestUnits.ValueInt64()),
+				"max_write_request_units": int(odts[0].MaxWriteRequestUnits.ValueInt64()),
+			}
+			action.OnDemandThroughput = expandOnDemandThroughput(v)
+		}
+	}
+
+	var wts []warmThroughputModel
+	response.Diagnostics.Append(data.WarmThroughputs.ElementsAs(ctx, &wts, false)...)
+	if len(wts) > 0 {
+		action.WarmThroughput = expandWarmThroughput(map[string]any{
+			"read_units_per_second":  int(wts[0].ReadUnitsPerSecond.ValueInt64()),
+			"write_units_per_second": int(wts[0].WriteUnitsPerSecond.ValueInt64()),
+		})
+	}
+
+	ut.GlobalSecondaryIndexUpdates = []ddbtypes.GlobalSecondaryIndexUpdate{
+		{
+			Create: action,
+		},
+	}
+
+	response.Diagnostics.Append(validateNewGSIAttributes(data, *table)...)
+
+	if response.Diagnostics.HasError() {
+		return
+	}
+
+	if utRes, err := conn.UpdateTable(ctx, ut); err != nil {
+		response.Diagnostics.AddError(
+			fmt.Sprintf(`Unable to create index "%s" on table "%s"`, data.Name.ValueString(), data.Table.ValueString()),
+			err.Error(),
+		)
+
+		return
+	} else {
+		for _, g := range utRes.TableDescription.GlobalSecondaryIndexes {
+			if aws.ToString(g.IndexName) == data.Name.ValueString() {
+				flattenGSI(ctx, *table, g, &data)
+
+				break
+			}
+		}
+	}
+
+	if _, err = waitTableActive(ctx, conn, data.Table.ValueString(), createTimeout); err != nil {
+		response.Diagnostics.AddError(
+			fmt.Sprintf(`Error while waiting for table "%s" to be active`, data.Table.ValueString()),
+			err.Error(),
+		)
+
+		return
+	}
```

Comment:
In Create, Update, and Delete, we call `UpdateTable`, then `waitTableActive`. This could be factored out.

For both Create and Update, we also call `waitGSIActive`. This could also be factored out.